### PR TITLE
Vision transformer self-supervised learning

### PR DESCRIPTION
**Pull request by daniel-j-h**
Experimenting with vision transformers and self-supervised learning in here.

To be specific: this changeset is implementing BYOL
- https://arxiv.org/abs/2006.07733
- https://www.youtube.com/watch?v=YPfUiOMYOEE

for self-supervised representation learning applied to the vision transformer architecture.

My main goals are to figure out if
- the vision transformer can be self-supervised with BYOL, too
- the vision transformer then learns good representations
- a downstream task then trains easier and faster compared to training from scratch (which is a bit of a pain)

cc @sandhawalia

### CODE REVIEW

**Comment by daniel-j-h** [comment]
Running this right now in self-supervised fashion on the open images v6 dataset.

What I'm seeing so far is it's collapsing as in different images result in the same constant embedding.

Ideas to experiment with
- Maybe our augmentations are not hard enough; look more closely how they do it in BYOL and SimCL
- Maybe with our transformers we don't need the projection heads and can work with the 768-dim directly
- Maybe it just doesn't work with transformers

Let's see what we can try here! :rocket: :moon:

**Comment by daniel-j-h** [comment]
I've always seen a collapse so far: for two different input images, we'd get the same outputs (model learns trivial `f(x)=c`).

After experimenting a few weeks it looks like I found a combination of augmentation, and model architecture, which at least does not immediately collapse. The changes so far are

Resize to a larger image before random cropping. Before we were using a `Resize(512)`, and then the random crops probably overlapped too much.

https://github.com/daniel-j-h/visiont-pytorch/blob/7f6e4b4a32eb00658d5efdc908e45351489b77f8/visiont/scripts/train.py#L96-L97

Apply more augmentations from the byol / simcl papers; latest torchvision makes that simple

https://github.com/daniel-j-h/visiont-pytorch/blob/7f6e4b4a32eb00658d5efdc908e45351489b77f8/visiont/scripts/train.py#L99-L101

No projection heads: in the BYOL paper they add projection heads to both architectures, but they don't explain why. Maybe because it worked for them. I've removed them here and instead of regressing on 256 dim tensors, we regress on the 768 dim tensors directly.

https://github.com/daniel-j-h/visiont-pytorch/blob/7f6e4b4a32eb00658d5efdc908e45351489b77f8/visiont/scripts/train.py#L119-L123

https://github.com/daniel-j-h/visiont-pytorch/blob/7f6e4b4a32eb00658d5efdc908e45351489b77f8/visiont/scripts/train.py#L137-L140

Using SGD right now, but kept the learning rate warmup (we used it with Adam) for now

https://github.com/daniel-j-h/visiont-pytorch/blob/7f6e4b4a32eb00658d5efdc908e45351489b77f8/visiont/scripts/train.py#L162

and this results in me not seeing a collapse so far! Which is good news! :tada: 

The next question is, are we actually learning good representations? And this we can only evaluate with a downstream task, or e.g. clustering the representations https://github.com/facebookresearch/faiss/ :v:

cc @sandhawalia

**Comment by sandhawalia** [comment]
❤️

**Comment by daniel-j-h** [comment]
Adding mixed precision training in https://github.com/daniel-j-h/visiont-pytorch/pull/1/commits/c65a8bc82be26fd847ae6a47775989aa5c45309a - latest pytorch makes it very simple to integrate :tada: On the 2x Titan RTX machine I'm training this on I'm seeing a reduction from almost seven hours down to three hours per epoch. Quite a lot; exciting! :racehorse: :dash:

**Comment by sandhawalia** [comment]
Houston we have a problem
